### PR TITLE
Fix EZP-24691: Move scope change to event so legacy-bridge can build …

### PR DESCRIPTION
**Pull request by wizhippo**
…the correct cli kernel

Closes #30

### CODE REVIEW

**Comment by andrerom** [comment]
@lolautruche Here is the LegacyBridge part of it I think ,as far as I know it replaces #30.

**Comment by lolautruche** [comment]
@andrerom I know, I'm already on it :wink:

File: bundle/EventListener/LegacyKernelListener.php [hunk 1]
```diff
@@ -43,6 +46,20 @@ function() {
 
     public function onConsoleCommand( ConsoleCommandEvent $event )
     {
+        $this->resetKernelHandler();
+
+        $eventDispatcher = $this->container->get( 'event_dispatcher' );
+        $eventDispatcher->addListener(MVCEvents::CONFIG_SCOPE_CHANGE, [$this, 'onConsoleConfigScopeChange'], -1);
+        $eventDispatcher->addListener(MVCEvents::CONFIG_SCOPE_RESTORE, [$this, 'onConsoleConfigScopeChange'], -1);
```
Comment (lolautruche) [inline bundle/EventListener/LegacyKernelListener.php]: Why adding those manually here? AFAICS, you could add those in `getSubscribedEvents()` right?

**Comment by emodric** [comment]
@wizhippo @lolautruche It works, but partially.

It fixes the issue where console cannot be ran at all, but clearing caches still has the same error:

```
Exception trace:
 () at /var/www/html/netgensite/ezpublish/cache/dev/ezpublishDevDebugProjectContainer.php:18929
 ezpublishDevDebugProjectContainer->getRequestService() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:327
 Symfony\Component\DependencyInjection\Container->get() at /var/www/html/netgensite/vendor/ezsystems/legacy-bridge/mvc/Kernel/Loader.php:157
 eZ\Publish\Core\MVC\Legacy\Kernel\Loader->eZ\Publish\Core\MVC\Legacy\Kernel\{closure}() at /var/www/html/netgensite/vendor/ezsystems/legacy-bridge/mvc/Kernel/Loader.php:117
 eZ\Publish\Core\MVC\Legacy\Kernel\Loader->eZ\Publish\Core\MVC\Legacy\Kernel\{closure}() at /var/www/html/netgensite/vendor/ezsystems/legacy-bridge/bundle/Cache/LegacyCachePurger.php:56
 eZ\Bundle\EzPublishLegacyBundle\Cache\LegacyCachePurger->getLegacyKernel() at /var/www/html/netgensite/vendor/ezsystems/legacy-bridge/bundle/Cache/LegacyCachePurger.php:66
 eZ\Bundle\EzPublishLegacyBundle\Cache\LegacyCachePurger->clear() at /var/www/html/netgensite/ezpublish/cache/dev/ezpublishDevDebugProjectContainer.php:50984
 eZBundleEzPublishLegacyBundleCacheLegacyCachePurger_0000000025cb8df6000000003b1f9169114e30a18e371fbe1096604cda598d0a->clear() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/CacheClearer/ChainCacheClearer.php:42
 Symfony\Component\HttpKernel\CacheClearer\ChainCacheClearer->clear() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php:70
 Symfony\Bundle\FrameworkBundle\Command\CacheClearCommand->execute() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:259
 Symfony\Component\Console\Command\Command->run() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:886
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:195
 Symfony\Component\Console\Application->doRun() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:96
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:126
 Symfony\Component\Console\Application->run() at /var/www/html/netgensite/ezpublish/console:27
```

File: bundle/EventListener/LegacyKernelListener.php (hunk 1 above)
Comment (wizhippo) [inline bundle/EventListener/LegacyKernelListener.php]: These only need to be attached when in console and not normailly.

**Comment by andrerom** [comment]
@wizhippo was the last change to fix what @emodric mentioned or something?

**Comment by wizhippo** [comment]
@andrerom yes the last fix has corrected the issue @emodric pointed out with the cache clearing issues.

**Comment by emodric** [comment]
@wizhippo @andrerom @lolautruche Cache clearing (and couple of other commands I tested) now work. Great job!

**Comment by emodric** [comment]
I do get an error when running `assets:install` command. I have no idea if it's related to this at all:

The `web` folder exists, ofcourse!

```
eddie@abyss: /var/www/html/netgensite $ php ezpublish/console assets:install --symlink --relative -vvv


                                              
  [InvalidArgumentException]                  
  The target directory "web" does not exist.  
                                              


Exception trace:
 () at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Command/AssetsInstallCommand.php:74
 Symfony\Bundle\FrameworkBundle\Command\AssetsInstallCommand->execute() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:259
 Symfony\Component\Console\Command\Command->run() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:886
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:195
 Symfony\Component\Console\Application->doRun() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:96
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /var/www/html/netgensite/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:126
 Symfony\Component\Console\Application->run() at /var/www/html/netgensite/ezpublish/console:27


assets:install [--symlink] [--relative] [--] [<target>]
```

**Comment by emodric** [comment]
#33 PR created which fixes the error in my comment above.

@wizhippo @lolautruche Can you do a quick glance at it please?

**Comment by emodric** [comment]
+1 for this :+1:

File: bundle/EventListener/LegacyKernelListener.php [hunk 2]
```diff
@@ -41,8 +44,22 @@ function() {
...
+    public function onConfigScopeChange( ScopeChangeEvent $event )
+    {
+        $this->resetKernelHandler();
+    }
+
     public function onConsoleCommand( ConsoleCommandEvent $event )
     {
+        $this->resetKernelHandler();
+
+        $eventDispatcher = $this->container->get( 'event_dispatcher' );
```
Comment (lolautruche) [inline bundle/EventListener/LegacyKernelListener.php]: Please inject the event dispatcher. I know we have the container, but always prefer injection over service locator.

**Comment by lolautruche** [comment]
+1 besides my last comment.

File: bundle/EventListener/LegacyKernelListener.php (hunk 2 above)
Comment (wizhippo) [inline bundle/EventListener/LegacyKernelListener.php]: Should I then just remove the containeraware and inject both the dispathcer and ezpublish_legacy.kernel_handler.cli?

Comment (wizhippo) [inline bundle/EventListener/LegacyKernelListener.php]: nevermind i misseted the use of it being uses to set the services

**Comment by lolautruche** [comment]
Thanks @wizhippo 

Review ping @andrerom @bdunogier @glye @yannickroger

**Comment by bdunogier** [comment]
Really clean, +1. Thank you.

File: bundle/EventListener/LegacyKernelListener.php [hunk 3]
```diff
@@ -15,13 +15,28 @@
...
+ * Resets legacy kernel handler when used in a command.
  */
 class LegacyKernelListener extends ContainerAware implements EventSubscriberInterface
 {
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+    function __construct( EventDispatcherInterface $eventDispatcher )
```
Comment (lolautruche) [inline bundle/EventListener/LegacyKernelListener.php]: Almost unsaw that. Missing `public` :wink:

**Comment by lolautruche** [comment]
OK @wizhippo . I just need the missing `public` operator before I merge this.

**Comment by wizhippo** [comment]
@lolautruche added `public` thanks.

**Comment by lolautruche** [comment]
Regression: if you try to launch a command and your instance is not installed/configured yet (e.g. no DB connection configured), you'll have a PDO exception:

```

  [Doctrine\DBAL\Driver\PDOException]
  SQLSTATE[HY000] [1045] Access denied for user 'root'@'localhost' (using password: NO)



Exception trace:
 () at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:47
 Doctrine\DBAL\Driver\PDOConnection->__construct() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOMySql/Driver.php:41
 Doctrine\DBAL\Driver\PDOMySql\Driver->connect() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:360
 Doctrine\DBAL\Connection->connect() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:429
 Doctrine\DBAL\Connection->getDatabasePlatformVersion() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:389
 Doctrine\DBAL\Connection->detectDatabasePlatform() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:328
 Doctrine\DBAL\Connection->getDatabasePlatform() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Persistence/Doctrine/DoctrineExpression.php:32
 eZ\Publish\Core\Persistence\Doctrine\DoctrineExpression->__construct() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Persistence/Doctrine/AbstractDoctrineQuery.php:72
 eZ\Publish\Core\Persistence\Doctrine\AbstractDoctrineQuery->__construct() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Persistence/Doctrine/ConnectionHandler.php:370
 eZ\Publish\Core\Persistence\Doctrine\ConnectionHandler->createSelectQuery() at /Users/lolautruche/workspace/ezsystems/ezpublish/ezpublish/cache/dev/ezpublishDevDebugProjectContainer.php:39153
 eZPublishCorePersistenceDoctrineConnectionHandler_0000000001b6b1b200000001634aec94c3ddff1eb3e43db8603f7487ea1756f5->createSelectQuery() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php:805
 eZ\Publish\Core\Persistence\Legacy\Content\Gateway\DoctrineDatabase->internalLoadContentInfo() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php:853
 eZ\Publish\Core\Persistence\Legacy\Content\Gateway\DoctrineDatabase->loadContentInfo() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Persistence/Legacy/Content/Gateway/ExceptionConversion.php:316
 eZ\Publish\Core\Persistence\Legacy\Content\Gateway\ExceptionConversion->loadContentInfo() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Persistence/Legacy/Content/TreeHandler.php:90
 eZ\Publish\Core\Persistence\Legacy\Content\TreeHandler->loadContentInfo() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Persistence/Legacy/Content/Handler.php:341
 eZ\Publish\Core\Persistence\Legacy\Content\Handler->loadContentInfo() at /Users/lolautruche/workspace/ezsystems/ezpublish/ezpublish/cache/dev/ezpublishDevDebugProjectContainer.php:36063
 eZPublishCorePersistenceLegacyContentHandler_0000000001b6b35300000001634aec94c3ddff1eb3e43db8603f7487ea1756f5->loadContentInfo() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Persistence/Cache/ContentHandler.php:91
 eZ\Publish\Core\Persistence\Cache\ContentHandler->loadContentInfo() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Repository/ContentService.php:380
 eZ\Publish\Core\Repository\ContentService->internalLoadContent() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/Repository/UserService.php:512
 eZ\Publish\Core\Repository\UserService->loadUser() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Bundle/EzPublishCoreBundle/ApiLoader/RepositoryFactory.php:85
 eZ\Bundle\EzPublishCoreBundle\ApiLoader\RepositoryFactory->buildRepository() at /Users/lolautruche/workspace/ezsystems/ezpublish/ezpublish/cache/dev/ezpublishDevDebugProjectContainer.php:2157
 ezpublishDevDebugProjectContainer->getEzpublish_Api_InnerRepositoryService() at /Users/lolautruche/workspace/ezsystems/ezpublish/ezpublish/cache/dev/ezpublishDevDebugProjectContainer.php:2148
 ezpublishDevDebugProjectContainer->{closure}() at /Users/lolautruche/workspace/ezsystems/ezpublish/ezpublish/cache/dev/ezpublishDevDebugProjectContainer.php:30391
 Closure->__invoke() at /Users/lolautruche/workspace/ezsystems/ezpublish/ezpublish/cache/dev/ezpublishDevDebugProjectContainer.php:30391
 eZPublishCoreRepositoryRepository_0000000001b6b4d500000001634aec94c3ddff1eb3e43db8603f7487ea1756f5->getURLAliasService() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/SignalSlot/Repository.php:373
 eZ\Publish\Core\SignalSlot\Repository->getURLAliasService() at /Users/lolautruche/workspace/ezsystems/ezpublish/ezpublish/cache/dev/ezpublishDevDebugProjectContainer.php:35443
 eZPublishCoreSignalSlotRepository_0000000001b6b4e500000001634aec94c3ddff1eb3e43db8603f7487ea1756f5->getURLAliasService() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Publish/Core/MVC/Symfony/Routing/Generator/UrlAliasGenerator.php:177
 eZ\Publish\Core\MVC\Symfony\Routing\Generator\UrlAliasGenerator->getPathPrefixByRootLocationId() at /Users/lolautruche/workspace/ezsystems/legacy-bridge/bundle/LegacyMapper/Configuration.php:281
 eZ\Bundle\EzPublishLegacyBundle\LegacyMapper\Configuration->getMultiSiteSettings() at /Users/lolautruche/workspace/ezsystems/legacy-bridge/bundle/LegacyMapper/Configuration.php:172
 eZ\Bundle\EzPublishLegacyBundle\LegacyMapper\Configuration->onBuildKernel() at n/a:n/a
 call_user_func() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/WrappedListener.php:61
 Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke() at n/a:n/a
 call_user_func() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:164
 Symfony\Component\EventDispatcher\EventDispatcher->doDispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:53
 Symfony\Component\EventDispatcher\EventDispatcher->dispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/ContainerAwareEventDispatcher.php:167
 Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/TraceableEventDispatcher.php:124
 Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch() at /Users/lolautruche/workspace/ezsystems/legacy-bridge/mvc/Kernel/Loader.php:215
 eZ\Publish\Core\MVC\Legacy\Kernel\Loader->eZ\Publish\Core\MVC\Legacy\Kernel\{closure}() at /Users/lolautruche/workspace/ezsystems/legacy-bridge/mvc/Kernel/Loader.php:114
 eZ\Publish\Core\MVC\Legacy\Kernel\Loader->eZ\Publish\Core\MVC\Legacy\Kernel\{closure}() at /Users/lolautruche/workspace/ezsystems/legacy-bridge/mvc/Kernel/Loader.php:320
 eZ\Publish\Core\MVC\Legacy\Kernel\Loader->resetKernel() at /Users/lolautruche/workspace/ezsystems/legacy-bridge/bundle/EventListener/ConfigScopeListener.php:38
 eZ\Bundle\EzPublishLegacyBundle\EventListener\ConfigScopeListener->onConfigScopeChange() at n/a:n/a
 call_user_func() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/WrappedListener.php:61
 Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke() at n/a:n/a
 call_user_func() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:164
 Symfony\Component\EventDispatcher\EventDispatcher->doDispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:53
 Symfony\Component\EventDispatcher\EventDispatcher->dispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/ContainerAwareEventDispatcher.php:167
 Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/TraceableEventDispatcher.php:124
 Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish-kernel/eZ/Bundle/EzPublishCoreBundle/EventListener/ConsoleCommandListener.php:67
 eZ\Bundle\EzPublishCoreBundle\EventListener\ConsoleCommandListener->onConsoleCommand() at n/a:n/a
 call_user_func() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/WrappedListener.php:61
 Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke() at n/a:n/a
 call_user_func() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:164
 Symfony\Component\EventDispatcher\EventDispatcher->doDispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:53
 Symfony\Component\EventDispatcher\EventDispatcher->dispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/ContainerAwareEventDispatcher.php:167
 Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/TraceableEventDispatcher.php:124
 Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:882
 Symfony\Component\Console\Application->doRunCommand() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:195
 Symfony\Component\Console\Application->doRun() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:96
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /Users/lolautruche/workspace/ezsystems/ezpublish/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:126
 Symfony\Component\Console\Application->run() at /Users/lolautruche/workspace/ezsystems/ezpublish/ezpublish/console:27
```

This is due to the fact that the legacy kernel loader loads the kernel service when calling `resetKernel()`, regardless it's already initialized or not. 
I'm currently working on a quick fix

**Comment by lolautruche** [comment]
PR open for the fix: #35

**Pull request merged**